app/controllers/admin/facilities_controller.rb

```diff
@@ -42,6 +42,24 @@ def destroy
     redirect_to admin_facilities_url, notice: 'Facility was successfully deleted.'
   end
 
+  def upload
+    authorize Facility
+    file = params[:upload_facilities_file]
+    @file_contents = ''
+    @errors = []
+    run_validations_and_read(file) if file.present?
+    if @errors.present?
+      @errors.prepend 'Please fix the errors below and try again:'
+      @errors << ['', 'Please contact team@simple.org if you need assistance']
+      flash.now[:alert] = @errors.join('<br/>').html_safe
+    elsif file.present?
+      facilities = Facility.parse_facilities(@file_contents)
+      ImportFacilitiesJob.perform_later(facilities)
+      flash.now[:notice] = 'File upload successful, your facilities will be created shortly.'
+    end
+    render :upload
```

Comment:
We should return a bad request if any of the params are missing, or the validations fail.